### PR TITLE
chore(tiering): Move cool entry warmup to DbSlice

### PR DESCRIPTION
**Pull request by dranikpg**
Simplify tiering code by moving warmup duty to db slice

### CODE REVIEW

File: src/core/compact_object.h [hunk 1]
```diff
@@ -375,22 +375,12 @@ class CompactObj {
...
-  template <typename T, typename... Args>
-  static std::enable_if_t<IsConstructibleFromArgs<T, Args...>, T*> AllocateMR(Args&&... args) {
+  template <typename T, typename... Args> static T* AllocateMR(Args&&... args) {
     void* ptr = memory_resource()->allocate(sizeof(T), alignof(T));
-    return new (ptr) T{std::forward<Args&&>(args)...};
+    if constexpr (std::is_constructible_v<T, decltype(memory_resource())> && sizeof...(args) == 0)
+      return new (ptr) T{memory_resource()};
+    else
+      return new (ptr) T{std::forward<Args>(args)...};
   }
```
Comment (dranikpg) [inline src/core/compact_object.h:384]: that's a trap, those two functions are not exlusive, to it all boils down to this if

File: src/server/tiered_storage.cc [hunk 2]
```diff
@@ -346,53 +341,22 @@ void TieredStorage::SetMemoryLowLimit(size_t mem_limit) {
...
-    decoder.ImportExternal(value);
-
-    auto cb = [readf = std::move(readf), decoder = std::move(decoder)](
-                  bool is_raw, const string* raw_val) mutable {
-      readf(DecodeString(is_raw, *raw_val, std::move(decoder)));
-      return false;
-    };
-    op_manager_->Enqueue(KeyRef(dbid, key), value.GetExternalSlice(), std::move(cb));
-  }
+  PrimeValue decoder;
```
Comment (romange) [inline src/server/tiered_storage.cc:351]: please keep dchecks verifying it's external, non cold item

File: src/server/tiered_storage.cc [hunk 3]
```diff
@@ -402,31 +366,19 @@ util::fb2::Future<T> TieredStorage::Modify(DbIndex dbid, std::string_view key,
...
-      if (is_raw) {
-        decoder.Materialize(*raw_val, true);
-        decoder.GetString(raw_val);
-      }
-      future.Resolve(modf(raw_val));
-      return true;
-    };
-    op_manager_->Enqueue(KeyRef(dbid, key), value.GetExternalSlice(), std::move(cb));
-  }
+  PrimeValue decoder;
```
Comment (romange) [inline src/server/tiered_storage.cc:369]: similarly, keep dcheck of not cool

File: src/server/tiered_storage.h [hunk 4]
```diff
@@ -78,8 +78,12 @@ class TieredStorage {
   // Run offloading loop until i/o device is loaded or all entries were traversed
   void RunOffloading(DbIndex dbid);
 
+  // Prune cool entries to reach the set memory goal with freed memory
   size_t ReclaimMemory(size_t goal);
 
+  // Returns the primary value, and deletes the cool item as well as its offloaded storage.
+  PrimeValue Warmup(DbIndex dbid, PrimeValue::CoolItem item);
```
Comment (romange) [inline src/server/tiered_storage.h:85]: please add it to non-linux stub as well.

File: src/server/tiered_storage.cc [hunk 5]
```diff
@@ -475,18 +427,10 @@ void TieredStorage::Delete(DbIndex dbid, PrimeValue* value) {
   DCHECK(value->IsExternal());
   ++stats_.total_deletes;
 
-  tiering::DiskSegment segment = value->GetExternalSlice();
```
Comment (romange) [inline src/server/tiered_storage.cc:478]: why this was removed?

File: src/server/tiered_storage.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/tiered_storage.cc:351]: true, lost them somehow

File: src/server/tiered_storage.cc (hunk 5 above)
Comment (dranikpg) [reply src/server/tiered_storage.cc:478]: Indeed, it can be triggered via a non command

File: src/server/tiered_storage.cc (hunk 3 above)
Comment (romange) [reply src/server/tiered_storage.cc:372]: it has not been resolved

File: src/server/tiered_storage.cc [hunk 6]
```diff
@@ -477,15 +432,12 @@ void TieredStorage::Delete(DbIndex dbid, PrimeValue* value) {
...
-    PrimeValue::CoolItem item = value->GetCool();
-    PrimeValue hot = DeleteCool(item.record);
-    DCHECK_EQ(OBJ_STRING, hot.ObjType());
+    auto hot = DeleteCool(value->GetCool().record);
+    DCHECK_EQ(hot.ObjType(), OBJ_STRING);
   }
 
   // In any case we delete the offloaded segment and reset the value.
   value->Reset();
-  stats_.total_deletes++;
```
Comment (romange) [inline src/server/tiered_storage.cc:488]: what about `stats_.total_deletes++` ?

**Review by romange — COMMENTED** [review]
Nice! Lots of code removal 👏🏼

File: src/server/tiered_storage.cc (hunk 6 above)
Comment (dranikpg) [reply src/server/tiered_storage.cc:488]: it's duplicated above

**Review by romange — APPROVED** [review]

**Pull request merged**